s3box_review/s3box.go

```diff
@@ -0,0 +1,243 @@
+	ErrInvalidJSONInput = fmt.Errorf("Only JSON-able inputs are supported for syncing to Redshift.")
+)
+
+// S3Box manages piping data into S3. The mechanics are to buffer data locally, ship to s3 when too much is buffered, and finally create manifests pointing to the data files.
+type S3Box struct {
+	// Inheret mutex locking/unlocking
+	sync.Mutex
+
+	// s3Bucket specifies the intermediary bucket before ultimately piping to Redshift. The user should have access to this bucket.
+	s3Bucket string
+
+	// s3Handler manages the piping of data to s3
+	s3Handler *s3.S3
+
+	// bufferSize is the maximum size of data we're willing to buffer before creating an s3 file
+	bufferSize int
+
+	// bufferedData is the data currently buffered in the box. Calling Dump ships this data into s3
+	bufferedData []byte
+
+	// timestamp tracks the time a box was created or reset
+	timestamp time.Time
+
+	// fileNumber indicates the number of s3 files which have currently been created
+	fileNumber int
+
+	// fileLocations stores the s3 files already created
+	fileLocations []string
+
+	// isSealed indicates whether writes are currently allows to the buffer
+	isSealed bool
+}
+
+// NewS3BoxOptions is the expected input for creating a new S3Box.
+// Currently only an S3Bucket is required. If AWS vars aren't explicitly provided, they'll
+// be pulled from your environment.
+type NewS3BoxOptions struct {
+	// S3Bucket is the destination s3 bucket.
+	// This is required.
+	S3Bucket string
+
+	// AWSKey is the AWS ACCESS KEY ID.
+	// By default grabs from your environment.
+	AWSKey string
+
+	// AWSPassword is the AWS SECRET ACCESS KEY.
+	// By default grabs from your environment.
+	AWSPassword string
+
+	// AWSToken is the AWS SESSION TOKEN.
+	// By default grabs from your environment.
+	AWSToken string
+
+	// BufferSize is the maximum size of data we buffer internally
+	// before creating an s3 file.
+	// This is optional and defaults to 10MB.
+	BufferSize int
+}
+
+// NewS3Box creates a new S3Box given the input options.
+// Errors occur if there's an invalid input or if there's difficulty setting up an s3 connection.
+func NewS3Box(options NewS3BoxOptions) (*S3Box, error) {
+	// Check for required inputs and a valid destination config
+	if options.S3Bucket == "" {
+		return nil, ErrS3BucketRequired
+	}
+
+	bufferSize := DefaultBufferSize
+	if options.BufferSize > 0 {
+		bufferSize = options.BufferSize
+	}
+
+	// Setup s3 handler and aws configuration. If no creds are explicitly provided, they'll be grabbed from the environment.
+	region, err := getRegionForBucket(options.S3Bucket)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get AWS region for bucket %s: (%s)", options.S3Bucket, err)
+	}
+
+	// If AWS creds were provided use those, otherwise grab them from your environment
+	var awsCreds *credentials.Credentials
+	if options.AWSKey == "" && options.AWSPassword == "" && options.AWSToken == "" {
```

Comment:
I feel like that behavior would be odd, because you could end up overriding some credentials and not others. Overall it feels unusual that someone would pass, say just an AWSKey or just an AWSPassword. If they try doing that, then the job will error and the user would figure out sooner than later what's going on.

I'm open to suggestions here in general, but my preference would be for there to be some type of erroring if it's configured incorrectly. My only thought would be a second check like 

```
else {
  if AWSKey == "" || AWSPassword == "" {
    return nil, fmt.Errorf("Error, missing key or Password")
  }
}
```